src/libcore/va_list.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation of a `va_list`.
+
+#[cfg(target_arch = "aarch64")]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+pub struct VaListImpl {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(any(target_arch = "arm", target_arch = "mips", target_arch = "powerpc64",
+          target_arch = "x86", all(target_arch = "x86_64", windows)))]
+/// Basic byte pointer variant of a `va_list`.
+pub type VaListImpl = *const i8;
```

Comment:
Can you put this one first? And maybe negate all the other special-cased architectures in this `#[cfg]`, so it's always defined? Otherwise `libcore` would stop compiling on the architectures not explicitly listed.

Reply:
Yeah

Reply:
After thinking about this more, I'm leaning toward failing to compile.

Reply:
This is unacceptable, IMO, since it will result in *libcore itself* not building for any architecture listed here. If you want to, you could instead make it so libcore builds but `VaList` is not defined. Or that the type is wholly incompatible with LLVM so trying to use the intrinsics fails in the LLVM verifier or something.

Reply:
For some reason I didn't think we needed to be able to compile libcore for an architecture not currently listed. If we do need to be able to compile libcore for other architectures I'll make the void pointer type the default.

Reply:
> If you want to, you could instead make it so libcore builds but VaList is not defined.

That would be my preference. Better to not have it defined (and get errors if you try to use it, and not pass the testsuite) than to define it incorrectly and have an obscure bug.

Reply:
Went back to using the void pointer variant as the default. I'm going to continue to look into doing something more clever, but in the meantime this will not cause libcore to fail to compile on architectures not explicitly listed.

Reply:
@dlrobertson Alright; hopefully tests will help catch any issues.